src/main/java/net/fabricmc/tinyremapper/util/VersionAwareMap.java

```diff
@@ -0,0 +1,373 @@
+        return getFeasibleVersionClasses(key.getName(), key.getVersion());
+    }
+
+    public Collection<ClassInstance> getFeasibleVersionClasses(String name, int version) {
+        return storage.get(name).stream()
+                .filter(cls -> cls.getVersionedName().getVersion() < version)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Return a single {@code ClassInstance} whose MRJ version is the largest
+     * among all candidates.
+     *
+     * <p> A candidate is any {@code ClassInstance} returned by method
+     * {@code getFeasibleVersions}. {@code null} will be returned if no
+     * candidate available. </p>
+     *
+     * @param key the name of {@code ClassInstance}
+     * @return a {@code ClassInstance} that is a candidate and MRJ version is the largest
+     */
+    public ClassInstance getByVersion(VersionedName key) {
+        return getByVersion(key.getName(), key.getVersion());
+    }
+
+    public ClassInstance getByVersion(String name, int version) {
+        return getFeasibleVersionClasses(name, version).stream()
+                .max(Comparator.comparingInt(o -> o.getVersionedName().getVersion()))
+                .orElse(null);
+    }
+
+    /**
+     * Associates the specified value with the specified key in this map
+     * (optional operation).  If the map previously contained a mapping for
+     * the key, the old value is replaced by the specified value.  (A map
+     * {@code m} is said to contain a mapping for a key {@code k} if and only
+     * if {@link #containsKey(Object) m.containsKey(k)} would return
+     * {@code true}.)
+     *
+     * @param key   key with which the specified value is to be associated
+     * @param value value to be associated with the specified key
+     * @return the previous value associated with {@code key}, or
+     * {@code null} if there was no mapping for {@code key}.
+     * (A {@code null} return can also indicate that the map
+     * previously associated {@code null} with {@code key},
+     * if the implementation supports {@code null} values.)
+     * @throws UnsupportedOperationException if the {@code put} operation
+     *                                       is not supported by this map
+     * @throws ClassCastException            if the class of the specified key or value
+     *                                       prevents it from being stored in this map
+     * @throws NullPointerException          if the specified key or value is null
+     *                                       and this map does not permit null keys or values
+     * @throws IllegalArgumentException      if some property of the specified key
+     *                                       or value prevents it from being stored in this map
+     */
+    @Override
+    public ClassInstance put(VersionedName key, ClassInstance value) {
+        ClassInstance result = remove(key);
+        storage.get(key.getName()).add(value);
```

Comment:
Need to prevent having duplicate versions in one entry array list.

Reply:
If a duplicate version exists, then it will be removed before put.

Reply:
I still think you should make private methods that removes locally given the array list. That way, you don't need to look into the storage map twice, for performance and cleanness concerns.

Reply:
Anything related to performance will be not be considered be this thing works. But I will definitely consider your suggestion when I do the final code clean-up.